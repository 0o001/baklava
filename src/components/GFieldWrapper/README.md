# GFieldWrapper

## Slots

| Name      | Description |
| --------- | ----------- |
| `default` | &nbsp;      |

## Props

| Name                  | Type       | Description | Default    |
| --------------------- | ---------- | ----------- | ---------- |
| `is-active-content`   | `Boolean`  |             | `false`    |
| `is-outline-label`    | `Boolean`  |             | `false`    |
| `is-value` *required* | `Boolean`  |             |            |
| `success`             | `Boolean`  |             | `false`    |
| `error`               | `Boolean`  |             | `false`    |
| `disable`             | `Boolean`  |             | `false`    |
| `label`               | `String`   |             | `""`       |
| `size`                | `String`   |             | `"big"`    |
| `feedback`            | `String`   |             | `""`       |
| `icon`                | `String`   |             | `""`       |
| `icon-position`       | `String`   |             | `""`       |
| `on-click`            | `Function` |             | `() => {}` |
| `on-outline-click`    | `Function` |             | `() => {}` |

## Data

| Name       | Type      | Description | Initial value |
| ---------- | --------- | ----------- | ------------- |
| `isActive` | `boolean` |             | `false`       |

## Computed Properties

| Name           | Type      | Description                                                                                                         |
| -------------- | --------- | ------------------------------------------------------------------------------------------------------------------- |
| `wrapperClass` | `object`  | **Dependencies:** `isOutlineLabel`, `isLabelTop`, `isActive`, `error`, `success`, `disable`, `iconPosition`, `size` |
| `iconColor`    | `unknown` | **Dependencies:** `error`, `disable`, `success`                                                                     |
| `isLabelTop`   | `unknown` | **Dependencies:** `isValue`                                                                                         |

## Events

| Name                     | Description                                                  |
| ------------------------ | ------------------------------------------------------------ |
| `update:isActiveContent` | <br/>**Arguments**<br/><ul><li>**`isActive: any`**</li></ul> |

## Methods

### clickWrapper()

**Syntax**

```typescript
clickWrapper(): void
```

### clickOutside()

**Syntax**

```typescript
clickOutside(): void
```

