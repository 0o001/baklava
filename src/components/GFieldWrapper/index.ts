import GFieldWrapper from './GFieldWrapper.vue';

export default GFieldWrapper;
