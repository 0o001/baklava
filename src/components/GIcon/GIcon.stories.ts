import { storiesOf } from '@storybook/vue';
import GIcon from './GIcon.vue';

storiesOf('GIcon', module)
  .addParameters({ component: GIcon })
  .add('Example', () => ({
    components: { GIcon },
    template: `<div> <GIcon name="alert-circle" strokeWidth='2px' /> <GIcon stroke="black" size="44px" fill="var(--orange-100)" name="star" /> </div>`,
  }))