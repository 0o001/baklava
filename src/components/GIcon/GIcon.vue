<template>
  <svg
    v-bind="this.attrs"
    v-on="$listeners"
    :class="classNames"
    :dataName="this.name"
    :dataTags="tags"
    :dataType="this.name"
    v-html="icon.contents"
  />
</template>

<script lang="ts">

import feather from 'feather-icons';

export default {
  name: 'GIcon',
  props: {
    animation: {
      type: String,
      default: undefined,
    },
    animationSpeed: {
      type: String,
      default: undefined,
    },
    fill: {
      type: String,
      default: 'none',
    },
    size: {
      type: [Number, String],
      default: '24px',
    },
    stroke: {
      type: String,
      default: 'currentColor',
    },
    strokeLinecap: {
      type: String,
      default: 'round',
    },
    strokeLinejoin: {
      type: String,
      default: 'round',
    },
    strokeWidth: {
      type: [Number, String],
      default: 2,
    },
    tag: {
      type: String,
      default: 'i',
    },
    name: {
      type: String,
      required: true,
      validator (name) {
        if (!feather.icons[name]) {
          throw new Error(`"${name}" is not an available icon type.`);
        }
        return true;
      },
    },
    color: {
      type: String,
      default: undefined,
    },
  },
  computed: {
    icon () {
      return feather.icons[this.name];
    },
    tags () {
      return feather.icons[this.name] ? feather.icons[this.name].tags : '';
    },
    classNames () {
      const classes = {
        'g-icon': true,
        feather: true,
        [`feather--${this.name}`]: true,
        [`feather--${this.animation}`]: this.animation,
        [`feather--${this.animationSpeed}`]: this.animationSpeed,
      };
      return classes;
    },
    attrs () {
      return {
        ...this.icon.attrs,
        fill: this.fill,
        height: this.size,
        stroke: this.color ? `var(--${this.color})` : this.stroke,
        'stroke-linecap': this.strokeLinecap,
        'stroke-linejoin': this.strokeLinejoin,
        'stroke-width': this.strokeWidth,
        width: this.size,
      };
    },
  },
};
</script>

<style lang="scss" scoped>
@keyframes feather--spin {
  from {
    transform: rotate(0);
  }
  to {
    transform: rotate(360deg);
  }
}
.feather {
  display: inline-block;
  overflow: hidden;
  &--spin {
    animation: feather--spin 2s linear infinite;
  }
  &--pulse {
    animation: feather--spin 2s infinite steps(8);
  }
  &--slow {
    animation-duration: 3s;
  }
  &--fast {
    animation-duration: 1s;
  }
  &__content {
    display: block;
    height: inherit;
    width: inherit;
  }
}
</style>
