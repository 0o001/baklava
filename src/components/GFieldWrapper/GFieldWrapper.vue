<template>
  <div>
    <div
      :class="wrapperClass"
      v-click-outside="clickOutside"
      @click="clickWrapper"
    >
      <label
        v-if="hasLabel"
        v-text="label"
      />
      <div class="content">
        <slot />
        <GIcon
          :class="iconPosition"
          v-if="icon"
          :name="icon"
          :color="iconColor"
        />
      </div>
      <span
        v-if="feedback"
        class="sub-label"
        v-text="feedback"
      />
    </div>
  </div>
</template>

<script lang="ts">

import ClickOutside from '../../directives/ClickOutside';
import GIcon from '../GIcon/GIcon.vue';

export default {
  name: 'GFieldWrapper',
  components: {
    GIcon,
  },
  directives: {
    ClickOutside,
  },
  props: {
    isActiveContent: {
      default: false,
      type: Boolean,
    },
    isOutlineLabel: {
      default: false,
      type: Boolean,
    },
    isValue: {
      required: true,
      type: Boolean,
    },
    success: {
      default: false,
      type: Boolean,
    },
    error: {
      default: false,
      type: Boolean,
    },
    disable: {
      default: false,
      type: Boolean,
    },
    label: {
      default: '',
      type: String,
    },
    size: {
      default: 'big',
      type: String,
    },
    feedback: {
      default: '',
      type: String,
    },
    icon: {
      default: '',
      type: String,
    },
    iconPosition: {
      default: '',
      type: String,
    },
    onClick: {
      default: () => {},
      type: Function,
    },
    onOutlineClick: {
      default: () => {},
      type: Function,
    },
  },
  data () {
    return {
      isActive: false,
    };
  },
  watch: {
    isActiveContent (newValue) {
      this.isActive = newValue;
    },
  },
  computed: {
    wrapperClass () {
      return {
        'g-field-wrapper': true,
        '-outline': this.isOutlineLabel,
        '-borderline': !this.isOutlineLabel,
        '-top': this.isLabelTop,
        '-active': this.isActive,
        '-error': this.error,
        '-valid': this.success,
        '-disable': this.disable,
        'icon-left': this.iconPosition,
        [this.size]: true,
      };
    },
    iconColor () {
      const color = this.error
        ? 'red-500'
        : this.disable
          ? 'mid-grey-500'
          : this.success
            ? 'green-500'
            : 'light-grey-500';
      return color;
    },
    isLabelTop: {
      get () {
        return this.isValue || false;
      },
    },
    hasLabel (): Boolean {
      return !!this.label.length;
    },
  },
  methods: {
    clickWrapper () {
      this.isActive = !this.isActive;
      this.$emit('update:isActiveContent', this.isActive);
      this.onClick();
    },
    clickOutside () {
      this.isActive = false;
      this.$emit('update:isActiveContent', this.isActive);
      this.onOutlineClick();
    },
  },
};
</script>

<style lang="scss" scoped>
.g-field-wrapper {
  position: relative;
  display: flex;
  flex-direction: column;

  label {
    z-index: 2;
    color: var(--dark-grey-500);
  }

  &.-outline {
    label {
      font-family: var(--font-family-base);
      font-size: var(--font-size-14);
      font-weight: 500;
      pointer-events: none;
      transition: 300ms ease all;
      margin-bottom: 5px;
      padding-left: 15px;
    }
  }

  &.-borderline {
    label {
      font-family: var(--font-family-base);
      font-size: var(--font-size-14);
      font-weight: 500;
      position: absolute;
      pointer-events: none;
      left: 15px;
      transition: 300ms ease all;
    }

    &.-top {
      label {
        top: -5px;
        font-size: var(--font-size-12);
        color: var(--orange-500);
        padding: 0;
        background: white;
        z-index: 2;
      }
    }
  }
  .sub-label {
    font-family: var(--font-family-base);
    font-size: var(--font-size-12);
    font-weight: normal;
    margin-top: 5px;
    margin-left: 15px;
    color: var(--dark-grey-500);
  }

  &.-disable {
    &.-borderline label {
      background: var(--bg-grey-500);
    }
  }

  /* Variants */

  &.-error {
    .content {
      border-color: var(--red-500) !important;
    }
    label,
    .sub-label {
      color: var(--red-500);
    }
  }
  &.-valid {
    .content {
      border-color: var(--green-500) !important;
    }
    label,
    .sub-label {
      color: var(--green-500);
    }
  }
  &.-disable {
    .content {
      border-color: var(--mid-grey-800) !important;
      input,
      textarea,
      select {
        background: var(--bg-grey-500);
        color: var(--main-grey-500);
      }
    }
    label,
    .sub-label {
      color: var(--dark-grey-500);
    }
  }

  .content {
    position: relative;
    border-radius: var(--radius-md);
    border-width: 1px;
    border-style: solid;
    border-color: var(--mid-grey-800);
    display: flex;
    width: 100%;

    input,
    textarea,
    select {
      font-family: var(--font-family-base);
      font-size: var(--font-size-14);
      width: 100%;
      display: block;
      padding: 11px 40px 11px 15px;
      background: none;
      border: none;
      border-radius: 5px;
      &:focus {
        outline: none;
      }
    }

    input[type='password'] {
      letter-spacing: 0.3em;
    }
    .g-icon {
      margin: 7px 10px 0;
      position: absolute;
      right: 0;

      &.icon-left{
        left: 0;
        margin-left: 15px;
      }
    }
  }
  &.big {
    .content {
      input,
      textarea,
      select {
        padding: 11px 40px 11px 15px;
      }
    }
    label {
      top: 13px;
    }
  }
  &.small {
    .content {
      input,
      textarea,
      select {
        padding: 7px 40px 7px 15px;
      }
    }
    label {
      top: 10px;
      font-size: var(--font-size-12);
    }
    .g-icon {
      width: 20px;
      margin-top: 3px;
    }
  }

  &.icon-left{
   .content{
     input,
     textarea,
     select {
       padding-left: 50px;
     }
   }
    label{
      left: 50px;
    }
  }
}
</style>
