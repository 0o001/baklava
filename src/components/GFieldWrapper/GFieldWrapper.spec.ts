import { mount, shallowMount, Wrapper } from '@vue/test-utils';
import GFieldWrapper from './GFieldWrapper.vue';
import GIcon from '../GIcon';

describe('GFieldWrapper', () => {
  let wrapper: Wrapper<any>;

  const renderWrapper = (options: any = {}) => {
    wrapper = shallowMount(GFieldWrapper, {
      ...options,
    });
  };

  it('should match snapshot when show is prop is given as false ', () => {
    renderWrapper({
      propsData: {
        isValue: false,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should props with default values ​​when no props are set', () => {
    renderWrapper({
      propsData: {
        isValue: false,
      },
    });
    expect(wrapper.vm.wrapperClass).toEqual({
      '-active': false,
      big: true,
      '-borderline': true,
      'icon-left': '',
      '-disable': false,
      '-error': false,
      '-outline': false,
      '-top': false,
      '-valid': false,
      'g-field-wrapper': true,
    });
    expect(wrapper.vm.iconColor).toEqual('light-grey-500');
  });

  it('should render given isValue when variant prop set', () => {
    renderWrapper({
      propsData: {
        isValue: true,
      },
    });
    expect(wrapper.vm.wrapperClass).toEqual({
      '-active': false,
      big: true,
      '-borderline': true,
      'icon-left': '',
      '-disable': false,
      '-error': false,
      '-outline': false,
      '-top': true,
      '-valid': false,
      'g-field-wrapper': true,
    });
  });

  it('should render given isOutlineLabel when variant prop set', () => {
    renderWrapper({
      propsData: {
        isValue: false,
        isOutlineLabel: true,
      },
    });
    expect(wrapper.vm.wrapperClass).toEqual({
      '-active': false,
      big: true,
      '-borderline': false,
      '-disable': false,
      '-error': false,
      '-outline': true,
      '-top': false,
      '-valid': false,
      'icon-left': '',
      'g-field-wrapper': true,
    });
  });

  it('should render given label and feedback when variant prop set', () => {
    renderWrapper({
      propsData: {
        isValue: false,
        label: 'TestLabel',
        feedback: 'TestSubLabel',
      },
    });
    expect(wrapper.find('label').text()).toEqual('TestLabel');
    expect(wrapper.find('.sub-label').text()).toEqual('TestSubLabel');
  });

  it('should not render feedback if feedback prop is not exists', () => {
    renderWrapper({
      propsData: {
        isValue: false,
        label: 'TestLabel',
      },
    });
    expect(wrapper.find('label').text()).toEqual('TestLabel');
    expect(wrapper.find('.sub-label').exists()).toEqual(false);
  });

  it('should render given success when variant prop set', () => {
    renderWrapper({
      propsData: {
        isValue: false,
        success: true,
      },
    });
    expect(wrapper.vm.wrapperClass).toEqual({
      '-active': false,
      big: true,
      '-borderline': true,
      'icon-left': '',
      '-disable': false,
      '-error': false,
      '-outline': false,
      '-top': false,
      '-valid': true,
      'g-field-wrapper': true,
    });
    expect(wrapper.vm.iconColor).toEqual('green-500');
  });

  it('should render given error when variant prop set', () => {
    renderWrapper({
      propsData: {
        isValue: false,
        error: true,
      },
    });
    expect(wrapper.vm.wrapperClass).toEqual({
      '-active': false,
      '-borderline': true,
      'icon-left': '',
      '-disable': false,
      '-error': true,
      '-outline': false,
      big: true,
      '-top': false,
      '-valid': false,
      'g-field-wrapper': true,
    });
    expect(wrapper.vm.iconColor).toEqual('red-500');
  });

  it('should render given disable when variant prop set', () => {
    renderWrapper({
      propsData: {
        isValue: false,
        disable: true,
      },
    });
    expect(wrapper.vm.wrapperClass).toEqual({
      '-active': false,
      '-borderline': true,
      'icon-left': '',
      '-disable': true,
      '-error': false,
      '-outline': false,
      big: true,
      '-top': false,
      '-valid': false,
      'g-field-wrapper': true,
    });
    expect(wrapper.vm.iconColor).toEqual('mid-grey-500');
  });

  it('should render given icon when variant prop set', () => {
    renderWrapper({
      propsData: {
        isValue: false,
        icon: 'info',
      },
    });
    expect(wrapper.findAllComponents(GIcon)).toHaveLength(1);
  });

  it('should class change when input is clicked', async () => {
    renderWrapper({
      propsData: {
        isValue: false,
      },
    });

    wrapper.find('.g-field-wrapper').trigger('click');
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.wrapperClass).toEqual({
      '-active': true,
      '-borderline': true,
      'icon-left': '',
      '-disable': false,
      '-error': false,
      big: true,
      '-outline': false,
      '-top': false,
      '-valid': false,
      'g-field-wrapper': true,
    });
  });

  it('should set active to false when input is clicked twice', async () => {
    renderWrapper({
      propsData: {
        isValue: false,
      },
    });

    wrapper.find('.g-field-wrapper').trigger('click');
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.wrapperClass).toEqual({
      '-active': true,
      '-borderline': true,
      'icon-left': '',
      '-disable': false,
      '-error': false,
      big: true,
      '-outline': false,
      '-top': false,
      '-valid': false,
      'g-field-wrapper': true,
    });

    wrapper.find('.g-field-wrapper').trigger('click');
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.wrapperClass).toEqual({
      '-active': false,
      '-borderline': true,
      'icon-left': '',
      '-disable': false,
      '-error': false,
      big: true,
      '-outline': false,
      '-top': false,
      '-valid': false,
      'g-field-wrapper': true,
    });
  });

  it('should render label when label prop set', async () => {
    renderWrapper({
      propsData: {
        isValue: false,
        label: 'dummy text',
      },
    });

    expect(wrapper.find('label').exists()).toBeTruthy();
    expect(wrapper.get('label').text()).toBe('dummy text');
  });

  it('should not render label when label prop is not set', async () => {
    renderWrapper({
      propsData: {
        isValue: false,
        label: '',
      },
    });

    expect(wrapper.find('label').exists()).toBeFalsy();
  });

  it('should render bottom-wrapper area if feedback is given', () => {
    // when
    renderWrapper({
      propsData: {
        isValue: false,
        feedback: 'feedback',
      },
    });

    // then
    expect(wrapper.find('.bottom-wrapper')).toMatchSnapshot();
  });

  it('should render bottom-wrapper area if bottom_right slot is given', () => {
    // when
    renderWrapper({
      slots: {
        bottom_right: 'bottom_right',
      },
      propsData: {
        isValue: false,
      },
    });

    // then
    expect(wrapper.find('.bottom-wrapper')).toMatchSnapshot();
  });
});
