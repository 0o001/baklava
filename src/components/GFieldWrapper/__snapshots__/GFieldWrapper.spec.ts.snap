// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GFieldWrapper should match snapshot when show is prop is given as false  1`] = `
<div>
  <div
    class="g-field-wrapper -borderline big"
  >
    <!---->
     
    <div
      class="content"
    >
       
      <!---->
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`GFieldWrapper should render bottom-wrapper area if bottom_right slot is given 1`] = `
<div class="bottom-wrapper g-d-justify-end">
  <!---->
  <div class="bottom-right-wrapper">bottom_right</div>
</div>
`;

exports[`GFieldWrapper should render bottom-wrapper area if feedback is given 1`] = `
<div class="bottom-wrapper g-d-justify-space-between"><span class="sub-label">feedback</span>
  <div class="bottom-right-wrapper"></div>
</div>
`;
